routers/web/repo/issue.go

```diff
@@ -580,56 +580,82 @@ func RetrieveRepoMilestonesAndAssignees(ctx *context.Context, repo *repo_model.R
-		return
-	}
-	projects2, err := db.Find[project_model.Project](ctx, project_model.SearchOptions{
-		ListOptions: db.ListOptionsAll,
-		OwnerID:     repo.OwnerID,
-		IsClosed:    util.OptionalBoolFalse,
-		Type:        repoOwnerType,
-	})
-	if err != nil {
-		ctx.ServerError("GetProjects", err)
-		return
-	}
+	if ctx.Repo.CanRead(unit.TypeProjects) {
+		projectsUnit, err := repo.GetUnit(ctx, unit.TypeProjects)
+		if err != nil {
+			ctx.ServerError("GetUnit", err)
+			return
+		}
+		projectsConfig := projectsUnit.ProjectsConfig()
 
-	ctx.Data["OpenProjects"] = append(projects, projects2...)
+		repoOwnerType := project_model.TypeIndividual
+		if repo.Owner.IsOrganization() {
+			repoOwnerType = project_model.TypeOrganization
+		}
 
-	projects, err = db.Find[project_model.Project](ctx, project_model.SearchOptions{
-		ListOptions: db.ListOptionsAll,
-		RepoID:      repo.ID,
-		IsClosed:    util.OptionalBoolTrue,
-		Type:        project_model.TypeRepository,
-	})
-	if err != nil {
-		ctx.ServerError("GetProjects", err)
-		return
-	}
-	projects2, err = db.Find[project_model.Project](ctx, project_model.SearchOptions{
-		ListOptions: db.ListOptionsAll,
-		OwnerID:     repo.OwnerID,
-		IsClosed:    util.OptionalBoolTrue,
-		Type:        repoOwnerType,
-	})
-	if err != nil {
-		ctx.ServerError("GetProjects", err)
-		return
-	}
+		var openProjects []*project_model.Project
+		if projectsConfig.ProjectsMode != repo_model.ProjectsModeOwner {
+			repoProjects, err := db.Find[project_model.Project](ctx, project_model.SearchOptions{
+				ListOptions: db.ListOptionsAll,
+				RepoID:      repo.ID,
+				IsClosed:    util.OptionalBoolFalse,
+				Type:        project_model.TypeRepository,
+			})
+			if err != nil {
+				ctx.ServerError("GetProjects", err)
+				return
+			}
+			openProjects = append(openProjects, repoProjects...)
+		}
+		if projectsConfig.ProjectsMode != repo_model.ProjectsModeRepo {
+			ownerProjects, err := db.Find[project_model.Project](ctx, project_model.SearchOptions{
+				ListOptions: db.ListOptionsAll,
+				OwnerID:     repo.OwnerID,
+				IsClosed:    util.OptionalBoolFalse,
+				Type:        repoOwnerType,
+			})
+			if err != nil {
+				ctx.ServerError("GetProjects", err)
+				return
+			}
 
-	ctx.Data["ClosedProjects"] = append(projects, projects2...)
+			openProjects = append(openProjects, ownerProjects...)
+		}
+
+		ctx.Data["OpenProjects"] = openProjects
+
+		var closedProjects []*project_model.Project
+		if projectsConfig.ProjectsMode != repo_model.ProjectsModeOwner {
+			repoProjects, err := db.Find[project_model.Project](ctx, project_model.SearchOptions{
+				ListOptions: db.ListOptionsAll,
+				RepoID:      repo.ID,
+				IsClosed:    util.OptionalBoolTrue,
+				Type:        project_model.TypeRepository,
+			})
+			if err != nil {
+				ctx.ServerError("GetProjects", err)
+				return
+			}
+
+			closedProjects = append(closedProjects, repoProjects...)
+		}
+		if projectsConfig.ProjectsMode != repo_model.ProjectsModeRepo {
```

Comment:
I am not quite a fan to handle the permission on this level.
It means that we have to add this check to any `list`/`find` route again.

Reply:
I moved permissions out of it and only kept projects mode check